### ISSUE TITLE
5k Output issues on iMac15,1 (27" 5k iMac - 2014)

### ISSUE THREAD

**Issue by khronokernel** [dortania/OpenCore-Legacy-Patcher#359]
When booting OpenCore on an iMac15,1, the iMac's firmware will enforce a 3840x2160 resolution on the internal display. This seems to be an unfortunate issue with the 5k iMac series as the issue plagued [Windows](https://www.reddit.com/r/mac/comments/3rxcih/windows_10_on_5k_imac_resolution_question/), [Linux](https://discussions.apple.com/thread/6626315) and [UEFI](https://askubuntu.com/questions/866920/limited-resolution-booting-with-refind-on-imac-5k). Apple eventually made a driver hack for Windows however both UEFI and Linux fixes are missing.

## Work-Arounds

~~The only current solution to restore 5k output is to boot macOS without any 3rd party boot loaders. However this therefore breaks OTA updates, AirPlay to Mac, Universal Control and other patches. Additionally forcing a higher resolution within macOS only emulates 5k, the actual screen resolution stays at 4k.~~

~~As we have no hardware locally, we cannot continue researching. Hardware donations would be appreciated, see [Supporting the patcher page](https://dortania.github.io/OpenCore-Legacy-Patcher/DONATE.html) for more information.~~

Thanks to the gracious donation from @turbomacs we have a 2014 5k iMac locally now! Currently research is being done and patch sets developed.

Working solution can be found here: https://github.com/dortania/OpenCore-Legacy-Patcher/issues/359#issuecomment-988417731

Associated branch: [5k patch](https://github.com/dortania/OpenCore-Legacy-Patcher/pull/777)

Issue will be closed once patch set is merged into mainline.

## Additional Information

Original Issues Thread:

* [Cannot correctly identify the iMac 5K resolution in nightly 0.1.7](https://github.com/dortania/OpenCore-Legacy-Patcher/issues/283)

**Comment by owenzhao** [comment]
Is that possible to just give the resolution options like 5K in 4K? In 5K, I was used to using the resolution of 2880 x 1620. There is no 2880x 1620 in 4K mode, the counter resolution was a little bigger about 3020x1698 as I remembered. However, the resolution bigger and smaller are the same, still 3200x1800 and 2560x1440. 

So I will be pleased if 2880x1620 could also be listed as an option under 4K. Is that possible? As I have almost never used the native 5K resolution, but use 2880x1620 a lot.

<img width="780" alt="截屏2021-07-05 上午7 42 33" src="https://user-images.githubusercontent.com/2182896/124402811-29f09480-dd65-11eb-9d63-a190702c09bf.png">

**Comment by jonblatho** [comment]
@owenzhao, I have tried this through SwitchResX and been unable to arrive at a fully correct solution. I have been able to get to 2880x1620, but not scaled at 2x (for the Retina display). Even then, since macOS believes the display is 4K, it still wouldn’t appear fully correct, particularly with aliased graphics, because macOS is still running downsampling for 4K resolution.

We know this because you actually *can* run at the display's native resolution of 5120x2880 (2560x1440@2x) through macOS’s display preferences and the 4K downsampling leads aliased graphics to look horrible.

**Comment by jonblatho** [comment]
I was poking around and stumbled on [a forum post](https://forums.macrumors.com/threads/scale-1680x1050-on-rmb-without-an-app.1873910/) about setting up a custom resolution on the 12" MacBook. Is there any chance there's a path to a fix through the `/System/Library/Displays/Contents/Resources/Overrides` directory?

**Comment by razordynamics** [comment]
Hi, I have this model and should be able to provide a hardware info dump with IOReg Explorer, will that help? Unfortunately I won't be able to donate actual hardware, since I'm using it currently, but I can provide info/testing/etc, whatever else might help.

**Comment by dgaastra** [comment]
Since Big Sur works with this iMac 15,1, would it not be possible for OpenCore to use parts of that driver?

**Comment by moritzj29** [comment]
Just to support the relevance of this issue, I would also love to see 5K support for the iMac15,1! It is the key feature of the product and the performance of the system is still very high. So I would like to use it for another couple of years :)

**Comment by lephitar** [comment]
Good to read so much enthusiasm for this issue. This iMac is also my workhorse for photography and I'd love to keep it running a few more years up to date

**Comment by sroby** [comment]
@dgaastra: As previously mentioned this is an issue at the firmware level, hacking the Big Sur driver into Monterey wouldn't solve it.

**Comment by ragian** [comment]
A lot of us uses the iMac 5K so an hadware donation is not possible, but what about a fundraiser? A second hand iMac 5k Late 2014 costs about 1000€ on ebay, we can raise the fund and sent to the guys (or raise it, buy the iMac and sent it).

**Comment by EverlastEngineering** [comment]
> Like I said previously, I am willing to donate a 2014 iMac if the team is ok with the display glass being broken/cracked but otherwise working. You would need to throw some clear packing tape over the cracked areas to prevent cuts but it would give you a machine to test on. Is that acceptable?

Doing god's work right now. Props man. Hopefully the devs can take it and find a solution!

I have a question about the symptom: is the main issue that there is bad aliasing at 4K?

**Comment by khronokernel** [comment]
@turbomacs Hi, thank you so much for offering, really appreciate it! So long as the iMac is functional with respect to the the motherboard, screen as well as a stock drive, should be no issue.

* Stock SSD/drive is needed to ensure firmware updates are allowed by macOS, as some 3rd party models will error out. Need to make sure that we have the latest firmware for testing. iMac15,1 uses PCIe based storage so it can be fussy about having an Apple made drive that's using the AHCI protocol.

My main concern is shipping, as sending a cracked iMac to Alberta, Canada would be quite costly as well as risking additional damages. Especially if you're located in the United States or even Europe.

Without doxxing your location, etc in the public issue, you can email me: khronokernel@icloud.com

* If I don't reply soon-ish after, just mention here. May have gotten into spam

**Comment by turbomacs** [comment]
@khronokernel I sent you an email.

**Comment by EverlastEngineering** [comment]
We're all here with fingers crossed. 😆

**Comment by wesmosis** [comment]
That sucks, I hope there is a solution since I have late iMac 5k Late 2014

**Comment by dgaastra** [comment]
Just checking in to see what the status is with our Alberta Magic. My wife is also from there.

Thanks
Dennis

**Comment by Jaaap** [comment]
According to https://www.reddit.com/r/MontereyPatcher/comments/nz1896/working_great_on_imac_5k_late_2014/ it broke in beta8.
Can we use the beta7 driver / firmware / files?

**Comment by deadsunrise** [comment]
> According to https://www.reddit.com/r/MontereyPatcher/comments/nz1896/working_great_on_imac_5k_late_2014/ it broke in beta8. Can we use the beta7 driver / firmware / files?

That's me. Here's a couple of screenshots, the resolution is not completely correct but at least it works, beta 8 broke everything for me and I haven't tried to upgrade yet.

<img width="1246" alt="CleanShot 2021-11-15 at 18 04 58@2x" src="https://user-images.githubusercontent.com/11164/141824186-626fbf41-fa9c-484d-a4c3-24b309cb70ba.png">
<img width="1251" alt="CleanShot 2021-11-15 at 18 04 03@2x" src="https://user-images.githubusercontent.com/11164/141824205-2edc007e-ead5-44cc-8f09-ca5997a8adaf.png">

**Comment by Jaaap** [comment]
Wait so does real 5k (5120 × 2880) work on beta6/7?

**Comment by khronokernel** [comment]
@Jaaap As mentioned at the start of this thread, this 5k output issue is entirely firmware based and was not introduced with macOS Monterey. Someone could boot Yosemite, Big Sur, etc and still have their display downgraded to 4k. Some firmware protocol is not being initialized and thus breaking if boot.efi is not the first loader to be started.

* I assume deadsunrise's issue was due to an AGDP bug Apple introduced with Beta 7/8 that affected native Macs as well. Beta 10+ should have resolved this, but off only restoring back to the previous state that we know currently.

turbomacs's has offered to send a machine so I respectfully ask everyone in this thread to be patient and wait till I receive the machine. I am as eager as everyone else to reverse and better understand the iMac's firmware structure with regards to the dual DP handling, however due to the sheer complexity of this issue there's no simple solution that can be done from macOS's kernel/userspace.

**Comment by deadsunrise** [comment]
I don't think so but I'm not sure. The thing is that it is detected as 3840x2160 and it does look very crisp and perfectly scaled to 2560x1440. I've been using this imac 5k for years and would say that it looks as good as always and I wouldn't notice that it's not detecting the correct resolution. 

<img width="781" alt="CleanShot 2021-11-15 at 18 26 59@2x" src="https://user-images.githubusercontent.com/11164/141826773-38f8810c-6a79-448f-8dfb-472281e1d723.png">
0

**Comment by turbomacs** [comment]
@khronokernel I have procured the following iMac to donate. It is being shipped to me now. I will inspect it, update it to the latest version of Big Sur and verify firmware’s are the latest.

Late 2014 iMac 27”
I5 3.5
8gb ddr3-1600
1tb fusion
2gb radeon 290

Display is not cracked however there is a line of dead pixels present. See photo. Please email me the shipping address and I will send the tracking number when it ships.

![E2B58B20-96C9-4074-8527-F8C87FFC0D55](https://user-images.githubusercontent.com/93228041/141878907-ac22bf7a-2657-4d42-bf20-922ddd98b561.jpeg)

**Comment by turbomacs** [comment]
Just an update. The donor iMac was delivered today. I didn’t get a chance to open it but I will tomorrow. Spirit of Jobs willing, I will ship it Friday for the long journey to Canada.

**Comment by turbomacs** [comment]
Another update: Updated to latest version of Big Sur. All firmware’s and boot roms updated to latest versions. DriveDX shows 90% life remaining on SSD portion of the fusion drive, so this machine has had light use.

Missed UPS pickup today but it will go out tomorrow.

<img width="592" alt="8B3B4F5B-8799-45D8-9369-C4738B6558D7" src="https://user-images.githubusercontent.com/93228041/142486495-554e5870-12eb-4c87-862c-6a807f7acc31.png">

**Comment by turbomacs** [comment]
Just letting everyone know this shipped yesterday. I hope it helps! Please be sure to update this thread if you manage to get it fixed so I remember. Good luck!

**Comment by sroby** [comment]
Since this is firmware level, are there any chances that fixing this will also solve the issue of not being able to access native resolution under Linux as well? That would be pretty awesome if that's the case.

**Comment by khronokernel** [comment]
> are there any chances that fixing this will also solve the issue of not being able to access native resolution under Linux as well?

Ideally yes! Assuming we can resolve this from UEFI land, all Ones should see a 5k display. Main concern however would be whether Linux' drivers will have a stroke seeing the dual DP output internally, but that's more of a concern for the Linux drivers to handle.

**Comment by mth75** [comment]
any news :-) ?

**Comment by dgaastra** [comment]
Please let us know the Paypal accounts for the procurement and development where we can donate after you guys get this thing working correctly on our iMac 15,1 5K. 
Thanks
Dennis

**Comment by Ausdauersportler** [comment]
Check this [link](https://dortania.github.io/OpenCore-Legacy-Patcher/DONATE.html#supporting-the-patcher)….

**Comment by khronokernel** [comment]
Hello everyone! Just wanted to pop into to say the iMac15,1 arrived saftely yesterday! 

Right now she's sitting on the kitchen table running tests, and I must say it's a beautiful machine. Even played around with AirPlay to Mac as a 4k Display for my development laptop.

And just want to say huge thank you again to @Turbomacs for sending the machine! It's an honor to be donated such a machine and get to work with this community!

Regarding researching, I will be posting mini blog posts in here about my findings. However please be respectful about asking for timelines, as neither I nor our developers can predict when or even if we'll be able to resolve this 5k bug. However I promise everyone here that I will put the utmost effort into this project.

- Mykola

Bonus pic of the system in action!
![IMG_4227](https://user-images.githubusercontent.com/48863253/144958557-4e3b3d63-4cb5-40d7-a6ca-be7759edf8e8.jpeg)

**Comment by khronokernel** [comment]
# iMac15,1 Research: Entry 1

So having the iMac for over a day, I was able to research and document some of the inital intricacies.

* [Affected Models](#affected-models)
* [Inital Observations: macOS](#initial-observations-macos)
  * [EDID Changes](#edid-changes)
  * [IOService Changes](#ioservice-changes)
* [Inital Observations: UEFI](#initial-observations-uefi)

## Affected Models

Unlike what was previously determined, it seems this 5k output bug affects all 5k iMacs potentially also including the iMac Pro.

* `iMac15,1`, `iMac17,1`, `iMac18,3`, `iMac19,1`, `iMac20,1`, `iMac20,2`, `iMacPro1,1`

We originally believed this issue to be isolated to 2014-2016 iMacs (15,1/17,1) however the issue was masked when Apple/AMD added 5k support to the Windows driver. Thus there are far less reports of output issues as well as complaints filed to Apple regarding this output issue.

## Inital Observations: macOS

Once I got the machine in hand, my first set of tests were to dump system nformation from both a clean slate as well as with OpenCore booted to better see how macOS handles every case.

From initial use, I could clearly see that the system had a much higher scaling option set just from the login window. Signing in we see that 1080p HiDPI is set and About This Mac reports 3840x2160 resolution. However System Profiler reports 5k regardless, likely some hard coded resolution until the dynamic reporting elsewhere.

### EDID Changes

After this, I started to probe EDID and found some interesting information. Namely the system changes Display Models and EDID outright when booting with OpenCore:

| Display  | Resolution | Model              | Vendor ID | Product ID | Supported Timings                       |
| :------- | :--------- | :----------------- | :-------- | :--------- | :-------------------------------------- |
| Stock    | 5120x2880  | APP Model ae**03** | 1552      | 4454**7**  | Missing from EDID (labeled dummy block) |
| OpenCore | 3840x2160  | APP Model ae**01** | 1552      | 4454**5**  | 3840x2160, 3200x1800, 2560x1440         |

Here we can see the fall back 4k mode has a more complete EDID, and seems Apple wanted the 4k mode to be a complete compatibility layer to avoid any issues in Ones they don't have full control over. Additionally because of this EDID change, some machines may experience broken scaling as macOS's DisplayOverrides profiles do not match the Product ID expected.

### IOService Changes

Looking into IOService, we can see Apple's display magic in effect.

On the GFX0 device (representing the R9 M290X), there are 4 total framebuffers present:

```
ATY,Namako@0 - Internal LCD     - LVDS 1.2 (0x2)
ATY,Namako@1 - Internal LCD     - LVDS 1.2 (0x2)
ATY,Namako@2 - External Display - DisplayPort 1.2 (0x400)
ATY,Namako@3 - External Display - DisplayPort 1.2 (0x400)
```

* LVDS can be seen as a version of DisplayPort for internal displays, commonly seen with laptops

Here we can see Apple is tunneling 2 DisplayPort 1.2 to help drive the 5120x2880 display (Model ID ae**03**)

However once we boot via OpenCore, IOService changes:

```
ATY,Namako@0 - Internal LCD     - LVDS 1.2 (0x2)
ATY,Namako@1 - Empty            - LVDS 1.2 (0x2)
ATY,Namako@2 - External Display - DisplayPort 1.2 (0x400)
ATY,Namako@3 - External Display - DisplayPort 1.2 (0x400)
```

We see that macOS only registers a single active DisplayPort 1.2 stream into a new 3840x2160 display (Model ID ae**01**).

## Inital Observations: UEFI

Doing a few tests, I was able to observe the following:

* On inital boot up and loading of the Apple Boot Picker, the display looks to be running at a 5120x2880 resolution
* When loading OpenCore, we see the screen blacks out shortly and returns as a 3840x2160 resolution
* Loading a UEFI Shell and immediately exiting resulted in the 4k output into macOS

We can see that some section of the iMac's firmware is detecting whether it's booting an unsupported entry, and will tell the Display Controller to force a compatibly mode with a new EDID and thus lock out the second DisplayPort/LVDS stream.

Because of how this lock out happens, we have 2 options:

* Look into whether we can communicate with the Display Controller to reset itself in macOS allowing the Dual DisplayPort stream
* Try to trick the firmware into thinking it's always booting macOS and no other 3rd party entries

## Closing Thoughts

From this initial testing batch, I've been able to learn quite a bit about how macOS and UEFI interacts with the system. Currently we still have many more tests to run in the coming weeks. Additionally we still need to research how Apple implemented the 5k Resolution fix in Windows back in 2016, as this will indicate to us how to best proceed.

**Comment by EverlastEngineering** [comment]
I love this. Following someone through the discovery and learning path is so utterly enjoyable, and one of my fav pastimes as well.

**Comment by viking1304** [comment]
@khronokernel I can confirm that most of the things you wrote are applicable to iMac18,3 too. Without OC I have DisplayProductID of ae03 and with OC I am getting ae11. There is also one additional EDID block if I boot without OC.

but.. there are 6 frame buffers instead of 4 that you see with your model:

ATY,Florin@0 - 0x2
ATY,Florin@1 - 0x2
ATY,Florin@2 - 0x400
ATY,Florin@3 - 0x400
ATY,Florin@4 - 0x400
ATY,Florin@5 - 0x400

I only know how to see if display-type string is set to NONE or to something else, so I can't give you more information unless you instruct me how to see if it is internal or external display. I can see that with OC ATY,Florin@0 have display-type of LCD, and the other 5 have NONE.

Hope this helps at least a little.

If you need more information from me, just let me know.

**Comment by macbookator** [comment]
I tried to use AirPlay to Mac with my 16" Intel MBP and iMac 27" 5K 2014 over Wi-Fi, but there is a resolution issue too. I can't choose resolution other than 1920x1080 and 1280x720 both on extended and mirrored, although iMac's screen resolution is 2560x1440. And it looks very bad.

Is it related to same issue, or I should create a new issue?

<img width="2560" alt="Ekran Resmi 2021-12-07 18 32 36" src="https://user-images.githubusercontent.com/94476177/145073790-c4049daf-a341-42eb-bde2-0791e03c226f.png">

<img width="836" alt="Ekran Resmi 2021-12-07 18 31 31" src="https://user-images.githubusercontent.com/94476177/145073989-5df4919f-55a5-4a36-a827-7e02c72e56c3.png">
Display Settings

![Ekran Resmi 2021-12-07 18 34 07](https://user-images.githubusercontent.com/94476177/145074025-e265b53e-a6c8-460b-a441-89c790d5e49d.png)
AirPlay Settings

**Comment by khronokernel** [comment]
@macbookator That is a bandwidth issue, make sure both machines are hooked up directly via Ethernet or Thunderbolt to enable 4k output. AirPlaySupport determines whether both the sender and receiver can handle a continuous 4k stream, if not it'll lock the option out to the end user. Using Wifi is not enough

With my setup, I have a mini Network Switch behind the iMac that I use to hook the iMac and MacBook Pro up directly allowing 4k output:
<img width="780" alt="Screen Shot 2021-12-05 at 2 48 18 PM" src="https://user-images.githubusercontent.com/48863253/145081306-78039df9-93e7-49c6-9a31-dfc839e9bd63.png">

**Comment by viking1304** [comment]
@macbookator I was typing almost the same message as @khronokernel, but he was faster. I already opened (and closed) this [ticket](https://github.com/dortania/OpenCore-Legacy-Patcher/issues/772) since I now have exactly the same resolutions as him over Thunderbolt.

**Comment by macbookator** [comment]
I didn't try this way because I don't have Ethernet to Thunderbolt 3 or Thunderbolt 2 to Thunderbolt 3 cable. I'll try it soon. Thank you both.

**Comment by khronokernel** [comment]
# iMac15,1 Research: Entry 2
* TL;DR: We've found a work-around to the 5k output bug!

Now having the machine for 2 days, I've been able to test some methodologies to try and trick the firmware into thinking we're only ever booting Apple supported enviroments. 

Currently the main 2 "supported" UEFI enviroments would be `boot.efi` and `diags.efi`. The former being a loader for XNU, latter being the loader for Apple's Diagnostic Tests.

While I could go over the main failed attempts from today, I'll make this post semi short and sweet by showing the more sucessful entries.

## diags.efi testing

With inital tests, I loaded up an ESP with Apple's Hardware Diagnostic Test to see where Apple "blessed" us with 5k output here. As semi expected, we did get 5k.

After this, I replaced `/S*/L*/C*/.diagnostics/Drivers/HardwareDrivers/Fan.efi` with OpenCore's loader and had the iMac boot `diags.efi` once again. And once I started the tests, OpenCore was triggered and I could see OpenCore's menu was rendered in 5120x2880!

Then booting into Monterey, we still retained 5k!
![Screen Shot 2021-12-07 at 5 58 39 PM](https://user-images.githubusercontent.com/48863253/145131555-d7dd00ad-56b4-4677-be14-05658c699ab1.jpg)

## Closing Thoughts

Overall I'm quite amazed that we've found a solution already, even if a bit "hacky". I will still be doing more research to better understand why loading this way works and whether we can integrate this into OpenCore itself and remove the reliance on diags.efi itself.

And for users hoping for 5k AirPlay to Mac support, know that currently Apple's AirPlaySupport.framework maxes out at 4K HDR. Thus while the iMac in front of me is running at 5k in Monterey, connecting my MacBook Pro over ethernet unfortunately does not yield a 5k option.

Currently I do not know how this affects either Windows and Linux support, more research and testing will be needed. Additionally I will unlikely add this fix into mainline OCLP for a bit to ensure no additional issues occur under this setup. However I will most likely create a new branch for users to experiment with.

- Mykola

**Comment by mth75** [comment]
Amazing! Great work you guys!

**Comment by khronokernel** [comment]
# iMac15,1 Research: First Public Test

So after cleaning up the patch set a bit, we should be good to test our first public build. As expected, ensure you have a native OS you can return to in the event the build has unexpected issues.

Binary below is based on the [`5k-display` branch](https://github.com/dortania/OpenCore-Legacy-Patcher/pull/777) (commit [fd26627](https://github.com/dortania/OpenCore-Legacy-Patcher/pull/777/commits/fd2662717c925fc6e95daf883141755e35e0e427))

* https://github.com/dortania/OpenCore-Legacy-Patcher/suites/4586913950/artifacts/123640811

Note: Github artifacts require you to be signed in.

**Comment by ElreboCM** [comment]
Great Job! Is it possible to install Monterey with the patcher on a USB drive?

**Comment by owenzhao** [comment]
> # iMac15,1 Research: First Public Test
> So after cleaning up the patch set a bit, we should be good to test our first public build. As expected, ensure you have a native OS you can return to in the event the build has unexpected issues.
> 
> Binary below is based on the [`5k-display` branch](https://github.com/dortania/OpenCore-Legacy-Patcher/pull/777) (commit [fd26627](https://github.com/dortania/OpenCore-Legacy-Patcher/pull/777/commits/fd2662717c925fc6e95daf883141755e35e0e427))
> 
> * https://github.com/dortania/OpenCore-Legacy-Patcher/suites/4586913950/artifacts/123640811
> 
> Note: Github artifacts require you to be signed in to download.

I downloaded the binary at the link. Build and install the patcher in the internal SSD. However, after I reboot, nothing happened. It seemed that I was not installed anything.

I have debug verbose to true. Any idea?

I was hoping to test native resolution issue so I thought it was OK to use macOS 11.6.1.

iMac 5K 2014 Late, 11.6.1 (20G224)

Howe could I know if I have successfully installed this patcher?

**Comment by gunmall** [comment]
It's working perfect! So cool. Thank you so much for your work.
![Screenshot_01](https://user-images.githubusercontent.com/94776437/145178665-8ccf8440-8a63-4422-9346-ff87817a8008.jpg)
.

**Comment by SofteqDG** [comment]
@khronokernel,
Thanks a lot for your work! First public build with 5K support works like a charm! 
![iMac15,1_OCPL_Monterey21,1](https://user-images.githubusercontent.com/7339487/145191610-7b22dc65-1a5f-46bb-b88f-43f541630fa9.jpg)

No new issues was detected after a small test.
Before installing this test build i've cleaned nvram and removed everything from /EFI volume.

**Comment by owenzhao** [comment]
> @khronokernel, Thanks a lot for your work! First public build with 5K support works like a charm! ![iMac15,1_OCPL_Monterey21,1](https://user-images.githubusercontent.com/7339487/145191610-7b22dc65-1a5f-46bb-b88f-43f541630fa9.jpg)
> 
> No new issues was detected after a small test. Before installing this test build i've cleaned nvram and removed everything from /EFI volume.

What did you do? After I booted and chose EFI, all I got was a blank black screen. I just install the patches to SSD EFI and chose booted from EFI. Any more I should do?

**Comment by SofteqDG** [comment]
@owenzhao,
I don't use picker. Update process for me was:
1. Change back resolution of the display to "Default for Display"
2. Clean nvram with sudo nvram -c in Terminal
3. Remove everything from EFI partition (with hidden files)
4. Install new public test build of OCPL
5. Turn off and turn on iMac
6. Be happy with 5K resolution :)

**Comment by viking1304** [comment]
I have a strange issue with this test build on iMac18,3

After I select that I want to boot from EFI

![image](https://user-images.githubusercontent.com/28989870/145195666-edbeaf9e-6d68-4159-8a1f-aa7638647a39.png)

I am getting this

![image](https://user-images.githubusercontent.com/28989870/145195830-188536dd-bf6b-4d71-9800-0ceb7588e7d2.png)

If I try to boot EFI I am getting blank black screen (even with -v boot argument in config). I will try again later after I finish with work.

@owenzhao since you have similar issues, this might be related to specific iMac model. What is your exact model?

**Comment by herrfisk** [comment]
Is this in the official build yet or is it in beta? I notice that the issue is still present on https://dortania.github.io/OpenCore-Legacy-Patcher/MODELS.html#imac

**Comment by owenzhao** [comment]
> I have a strange issue with this test build on iMac18,3
> 
> After I select that I want to boot from EFI
> 
> ![image](https://user-images.githubusercontent.com/28989870/145195666-edbeaf9e-6d68-4159-8a1f-aa7638647a39.png)
> 
> I am getting this
> 
> ![image](https://user-images.githubusercontent.com/28989870/145195830-188536dd-bf6b-4d71-9800-0ceb7588e7d2.png)
> 
> If I try to boot EFI I am getting blank black screen (even with -v boot argument in config). I will try again later after I finish with work.
> 
> @owenzhao since you have similar issues, this might be related to specific iMac model. What is your exact model?

I have got mine machine work now.

<img width="818" alt="截屏2021-12-08 下午7 26 00" src="https://user-images.githubusercontent.com/2182896/145203169-795dae5b-e6cf-47f4-b208-632d92a7d357.png">

## For the black screen issue. 

That was because of the second EFI, you should never chose that. You should first choose EFI Boot, then chose the installer or the macOS you wanted to use. If you chose EFI again, you got a black screen. 

If you don't want to use this patcher, you should choose the macOS menu along with EFI boot menu. So clearly the second EFI should never be chose and should not be shown as a menu at all.

Good luck!😊

**Comment by mth75** [comment]
> 3\. Remove everything from EFI partition (with hidden files)

How did you remove the EFI partition?

**Comment by ElreboCM** [comment]
I can confirm it is also working for me. 

However, there is one caveat with the method atm:
**Keyboard Layout is English although I have set it to German.**

**Comment by khronokernel** [comment]
# iMac15,1 Research: Public Test Build 2

Cleaned up the OCLP build process as well as ensure that no "mysterious" EFI BOOT entry appearing in OpenCore's menu for users to accidentally select.

Binary below is based on the [`5k-display` branch](https://github.com/dortania/OpenCore-Legacy-Patcher/pull/777) (commit f75f1aa)

* https://github.com/dortania/OpenCore-Legacy-Patcher/suites/4593775671/artifacts/123928482

**Comment by mth75** [comment]
> # iMac15,1 Research: Public Test Build 2
> Cleaned up the OCLP build process as well as ensure that no "mysterious" EFI BOOT entry appearing in OpenCore's menu for users to accidentally select.
> 
> Binary below is based on the [`5k-display` branch](https://github.com/dortania/OpenCore-Legacy-Patcher/pull/777) (commit [f75f1aa](https://github.com/dortania/OpenCore-Legacy-Patcher/commit/f75f1aa87a6718255e8172ebbafeb0a23c2249c2))
> 
> * https://github.com/dortania/OpenCore-Legacy-Patcher/suites/4593775671/artifacts/123928482
<img width="698" alt="Screenshot 2021-12-08 at 19 53 00" src="https://user-images.githubusercontent.com/46650192/145266878-31bc8999-5c77-4bb7-a6b2-f51c35b4aeeb.png">

Works on my: Thx mate! I was running Monterey already so I know :

a. Build Opencore again
b Installed Opencore to Internal drive (overwrote the EFI partition).
c. rebooted ...

done!

**Comment by khronokernel** [comment]
So after some extensive testing, can comfortably say this new digs.efi chain loading should be quite reliable. I have merged this into mainline earlier today:

* https://github.com/dortania/OpenCore-Legacy-Patcher/pull/777

We'll likely do a new release for OpenCore Legacy Patcher later this week once we have the remaining cruft cleaned up.

----

But otherwise we can finally consider this issue closed! I have summarizing my research and development of this thread on my blog for those interested:

* https://khronokernel.github.io/macos/2021/12/08/5K-UEFI.html

-----

Finally I wish to say thank you again to @turbomacs for offering to both send and cover shipping for this beautiful 5k iMac all the way to this frozen land of Alberta, Canada. Without their extremely generous donation, I don't believe I would have been able to get so far!

**Comment by herrfisk** [comment]
Remember to update the "Supported models" page that this is now fixed

**Comment by Caesar113** [comment]
I used the latest OpenCore release to upgrade my iMac 15,1 Big Sur installation to Monterey 12.1 last night and it went seamlessly! Many thanks!!!

**Comment by herrfisk** [comment]
> I'm trying to install the Monterey 12.1 but after a while a black screen with a folder with a question mark inside it is showed.
> Any advice?

Yes this happened to me as well. It’s because during reboots the computer boots from the regular, unpatched EFI. 
You need to boot and hold down the option key and choose the open patcher EFI and then your hard drive. The computer will restart a couple of times during the process so if you get stuck on this ore some other error you have probably booted to the unpatched EFI. Just reboot and hold down option
